### DISCUSSION TITLE
Is it possible to set breakpoints in SYCL kernel code in RelWithDebInfo configuration?

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1308] [Q&A] [answered]
Hello. To workaround my [issue](https://github.com/AdaptiveCpp/AdaptiveCpp/issues/1298) with Debug builds, I've tried to build binary using `RelWithDebInfo` configuration with the `-O0` compilation option and `omp.library-only` target. I encountered an issue where breakpoints set in SYCL kernel code are not being hit. Breakpoints set in the host code are hit as expected.

I have tried using both the `cpu_selector` and `host_selector`, but the issue persists in both cases.

Could the inability to hit breakpoints in the kernel code be due to using the `RelWithDebInfo` configuration instead of a full `Debug` configuration?

Is there a known limitation or issue with AdaptiveCpp regarding the debugging of SYCL kernels?

Are there specific steps or configurations required to enable debugging within SYCL kernel code that I might be missing?

Thank you in advance for your help and support.

**Comment by illuhad** [comment]
There are no known limitations regarding the ability to debug kernels in `omp.library-only` on Linux with regular host debuggers.

What kind of kernel are you looking at? `single_task`, `parallel_for` with `range` argument, `parallel_for` with `nd_range` argument but without barriers, or `parallel_for` with `nd_range` argument and with barriers?

* `single_task` is just straight C++ function call. I don't see how this could cause problems.
* `parallel_for` with `range` argument boils down to a `#pragma omp parallel for` loop. I don't know whether there are debugging limitations in OpenMP loops on Windows. If there are, they would also translate to our SYCL kernels.
* `parallel_for` with `nd_range` argument boils down to a `#pragma omp parallel for` loop across work groups, where `boost::fiber` objects are handling concurrency between work items. If there are no barriers, then only a single fiber is constructed and will execute all of the code in a work group. If the program flow hits a SYCL barrier, it assigns one fiber per work item and cycles through them to ensure all work items can hit the barrier at the same time. Fiber objects have their own stack that is separate from the regular stack, and cycling between them additionally requires switching stacks. I don't know how `boost::fiber` implements this on Windows, and to what extent the mechanisms used there affect debugging.

**Comment by blinkfrog** [comment]
Thanks. It's just a basic parallel kernel using `parallel_for` with `range` argument. Probably, it's `RelWithDebInfo` that prevents debugging of kernels - may be OpenMP checks if NDEBUG is defined to achieve a proper parallel loops debugging, and it is defined in RelWithDebInfo. So, basically, I need to fix my initial problem first.

**Comment by blinkfrog** [comment]
Just want to add that it is only cpu/cpu.accelerated devices, in which kernels breakpoints don't hit. When working with NVidia devices and debugger, breakpoints hit just fine in the kernel code.

**Comment by blinkfrog** [comment, marked as answer]
Breakpoints hit in kernels on cpu with `RelWithDebInfo` - just need to set more aggressive debug compile options, like these:

```
        -g
        -Og           # or better -O0, but compilation sometimes hangs in big projects
        -fno-inline
        -fno-omit-frame-pointer
        -fno-limit-debug-info
        -fstandalone-debug
        -fextend-variable-liveness=all
```
